compiler/fir/tree/tree-generator/src/org/jetbrains/kotlin/fir/tree/generator/FirTreeBuilder.kt

```diff
@@ -108,6 +108,7 @@ object FirTreeBuilder : AbstractFirTreeBuilder() {
     val callableReferenceAccess = element("CallableReferenceAccess", Expression, qualifiedAccessExpression)
     val thisReceiverExpression = element("ThisReceiverExpression", Expression, qualifiedAccessExpression)
     val expressionWithSmartcast = element("ExpressionWithSmartcast", Expression, qualifiedAccessExpression)
+    val expressionWithSmartcastToNull = element("ExpressionWithSmartcastToNull", Expression, qualifiedAccessExpression)
```

Comment:
I thought of that too, and decided to make it one-of-a-kind in order to avoid inception that this is kind of smartcast. Well, it's kind of smartcast, but not in the same way: no materialization during conversion to backend IR, not all smartcast types are intersected. So, during its introduction, I'd like to visit every use of existing smartcast expression kind, and make sure this new kind of expression is treated properly per situation.

Well, now every use/override point is visited, maybe we can restructure that on top of the current state. @demiurg906 WDYT?

Reply:
Originally I thought that `FirExpressionWithSmartcastToNull` will fully delegate to original expression and won't have any smartcast type. But after review I see that there may be situations where there are situations when expression has meaningful smartcasts and `Nothing?` in same time. So current approach that `FirExpressionWithSmartcastToNull` is independent from `FirExpressionWithSmartcast` but fully copy it is not good. I see two options here:
1. Take Marks suggestion and inherit `FirExpressionWithSmartcastToNull` from `FirExpressionWithSmartcast` (since it adds new semantic to existed one)
2. Make `FirExpressionWithSmartcastToNull` dummy delegate to original expression without any `smartcasted` types and in case when expression has e.g. smartcast to `A?` and `Nothing?` wrap it twice: firstly in `FirExpressionWithSmartcast` and then in `FirExpressionWithSmartcastToNull`

I prefer option 1.